scanpy/tools/_rank_genes_groups.py

```diff
@@ -341,12 +341,14 @@ def logreg(self, **kwds):
         clf = LogisticRegression(**kwds)
         clf.fit(X, self.grouping.cat.codes)
         scores_all = clf.coef_
-        for igroup, _ in enumerate(self.groups_order):
+        unique_codes = np.unique(self.grouping.cat.codes)
+        for igroup, cat in enumerate(self.groups_order):
             if len(self.groups_order) <= 2:  # binary logistic regression
                 scores = scores_all[0]
             else:
-                scores = scores_all[igroup]
-
+                cat = np.where(self.grouping.cat.categories == cat)[0][0]
+                cat = np.where(unique_codes == cat)[0][0]
```

Comment:
can you please walk us through the logic here?

what’s the goal, what does each step do?

Reply:
There is one issue with this solution. `self.grouping.cat.categories` still has all categories. So we get an error. I updated the test to account for this behaviour. The previous solution does account for this.

Reply:
Great! I added some comments, now it’s ready.